src/modules/hunting/etcd.py

```diff
@@ -0,0 +1,142 @@
+class etcdRemoteAccessActive(ActiveHunter):
+    """Checks for remote write access to etcd"""
+
+    def __init__(self, event):
+        self.event = event
+
+    def db_keys_write_access(self):
+        logging.debug(self.event.host)
+        logging.debug("Active hunter is attempting to write keys remotely")
+        data = {
+            'value': 'remote write access penetration'
+        }
+        r_secure = "https://{host}:{port}/v2/keys/message".format(host=self.event.host, port=2379)
+        r_not_secure = "https://{host}:{port}/v2/keys/message".format(host=self.event.host, port=2379)
+
+        res = helperFuncDo2Requests(r_secure, r_not_secure)
+        if res:
+            self.publish_event(etcdRemoteWriteAccessEvent())
+            return True
+        return False
+
+    def execute(self):
+        self.db_keys_write_access()
+
+@handler.subscribe(OpenPortEvent, predicate=lambda p: p.port == 2379)
+class etcdRemoteAccess(Hunter):
+    """Etcd Remote Access
+    Checks for remote availability of etcd, version, read access, write access
+    """
+    # TODO:
+    # Check the etcd hunter on a remote cluster! (currently everything was checked only at 127.0.0.1:2379)
+    def __init__(self, event):
+        self.event = event
+
+    def db_keys_disclosure(self):
+        logging.debug(self.event.host)
+        logging.debug("Passive hunter is attempting to read etcd keys remotely")
+        r_secure = "https://{host}:{port}/v2/keys".format(host=self.event.host, port=2379)
+        r_not_secure = "http://{host}:{port}/v2/keys".format(host=self.event.host, port=2379)
+        res = helperFuncDo2Requests(r_secure, r_not_secure)
+        if res:
+            self.publish_event(etcdRemoteReadAccessEvent(res.content))
+            return True
+        return False
+
+    def version_disclosure(self):
+        logging.debug(self.event.host)
+        logging.debug("Passive hunter is attempting to check etcd version remotely")
+        r_secure = "https://{host}:{port}/version".format(host=self.event.host, port=2379)
+        r_not_secure = "http://{host}:{port}/version".format(host=self.event.host, port=2379)
+        res = helperFuncDo2Requests(r_secure, r_not_secure)
+        if res:
+            self.no_auth_evidence = res.content
+            self.publish_event(etcdRemoteVersionDisclosureEvent(res.content))
+            return True
+        return False
+
+    def unauthorized_access(self):
+        logging.debug(self.event.host)
+        logging.debug("Passive hunter is attempting to access etcd without authorization")
+        r_not_secure = "http://{host}:{port}/version".format(host=self.event.host, port=2379)
+        res = helperFuncDo2Requests(r_not_secure, r_not_secure)# We dont have to do 2 requests this time
+        if res:
+            self.publish_event(etcdAccessEnabledWithoutAuthEvent(res.content))
+            return True
+        return False
+
+    def execute(self):
+        if self.version_disclosure():
```

Comment:
Another thing to consider would be using the version disclosure test to publish two separate events: one for secure access, one for not-secure access. Then you could have something along these lines:

```python
@handler.subscribe(EtcdInsecureAccessEvent)
class EtcdRemoteAccess(Hunter):
    def __init__(self, event):
        self.event = event
        self.protocol = "http"

    def db_keys_disclosure(self):
        r = requests.get("{protocol}://{host}:{port}/v2/keys".format(protocol=self.protocol, host=self.event.host, port=2379))
        self.content = r.content
        return (r.status_code == 200 and r.content != "")

    def execute(self):
        if self.db_keys_disclosure():
             self.publish_event(EtcdRemoteReadAccessEvent(self.content))

@handler.subscribe(EtcdSecureAccessEvent)
class EtcdSecureRemoteAccess(EtcdRemoteAccess):
    def __init__(self, event):
        self.event = event
        self.protocol = "https://"
```

The Insecure version would run if we established that we could access on http://, and the Secure version would run if we established that we could access on https://,  inheriting all the same tests as the Insecure one, but would use the different protocol string. You then wouldn't need the helper functions as the same logic would be taken care of by the classing, which I think is a bit nicer. 

We could also have another prarameter on the events like EtcdRemoteReadAccessEvent to indicate whether it was the secure or insecure version

Reply:
Are you sure?
Maybe I just prefer the current version just because I don't like splitting the code into different classes and methods; But I think it would make the code harder to understand since we would have 2 classes for the subscribers (instead of 1) that both publishes the same events (since we don't distinguish between an insecure- HTTP  and an secure- HTTPS events).
and 2 more events ( EtcdSecureAccessEvent, EtcdInsecureAccessEvent).

That is true we wouldn't have the Helper functions, but I think they are intuitive and simple to understand.